community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java

```diff
@@ -0,0 +1,163 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.countStore;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.neo4j.kernel.impl.store.counts.keys.CountsKey;
+import org.neo4j.kernel.impl.store.counts.keys.CountsKeyType;
+import org.neo4j.kernel.impl.store.counts.keys.IndexSampleKey;
+import org.neo4j.kernel.impl.store.counts.keys.IndexStatisticsKey;
+import org.neo4j.kernel.impl.store.counts.keys.NodeKey;
+import org.neo4j.kernel.impl.store.counts.keys.RelationshipKey;
+import org.neo4j.kernel.impl.store.kvstore.UnknownKey;
+import org.neo4j.kernel.impl.transaction.log.ReadableLogChannel;
+import org.neo4j.kernel.impl.transaction.log.WritableLogChannel;
+
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyFactory.indexSampleKey;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyFactory.indexStatisticsKey;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyFactory.nodeKey;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyFactory.relationshipKey;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.ENTITY_NODE;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.ENTITY_RELATIONSHIP;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.INDEX_SAMPLE;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.INDEX_STATISTICS;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.value;
+
+public class Snapshot
+{
+    private ConcurrentHashMap<CountsKey,long[]> map;
+    private long txId;
+
+    public Snapshot( long txId, ConcurrentHashMap<CountsKey,long[]> map )
+    {
+        this.map = map;
+        this.txId = txId;
+    }
+
+    public ConcurrentHashMap<CountsKey,long[]> getMap()
+    {
+        return map;
+    }
+
+    public long getTxId()
+    {
+        return txId;
+    }
+
+    public static void serialize( WritableLogChannel channel, Snapshot snapshot ) throws UnknownKey, IOException
+    {
+        channel.putLong( snapshot.getTxId() );
+        channel.putInt( snapshot.getMap().size() );
+
+        for ( Map.Entry<CountsKey,long[]> pair : snapshot.getMap().entrySet() )
+        {
+            CountsKey key = pair.getKey();
+            long[] value = pair.getValue();
+
+            switch ( key.recordType() )
+            {
+
+            case ENTITY_NODE:
+                NodeKey nodeKey = (NodeKey) key;
+                channel.put( ENTITY_NODE.code );
+                channel.putInt( nodeKey.getLabelId() );
+                channel.putLong( value[0] );
```

Comment:
we can add `assert value.length == 1` here and corresponding assertions for the other cases as well